### DISCUSSION TITLE
How to improve performance?

### DISCUSSION THREAD

**Discussion by patrikpatrik** [DanielChappuis/reactphysics3d#307] [General]
Hello again Daniel,

Just curious, is there any way to increase performance ? For instance is there a way to send all this to the compute shader for the collision detection & resolution to run parallel?

I've been doing some benchmark testing and came to the conclusion that I can't have any more than 200-300 objects without slowing down considerably. This was tested with the default cube, default sphere, default capsule, a custom convex cube, custom convex 3-sided equilateral triangle, and a 4-sided pyramid. Also debug lines and triangles set to `off` and sleep `enabled`.

My hypothesis was that the 3-sided equilateral triangle with 4 vertices and only 4 sides would be less calculation intensive, but that wasn't the case at all.

**Comment by DanielChappuis** [comment]
First, make sure that you are compiling the library in **Release** mode and not in **Debug** mode. If you want to obtain more performance, you can try to decrease the number of iterations of the velocity and position solver (as described [here](https://reactphysics3d.com/usermanual.html#x1-210008.1.1)). Note that it will decrease the quality of the physics simulation.

In the future, I will make possible to use multi-threading to run the physics simulation in parallel on several CPUs but this is not possible at the moment.

**Reply by patrikpatrik** [reply]
@DanielChappuis, thanks for the tip. And lastly, what about data structures like octtrees. Would that be possible and would it vastly improve performance as well?

**Reply by DanielChappuis** [reply]
The broad-phase collision detection of ReactPhysics3D is already implemented using a Dynamic AABB Tree which gives the same kind of performance improvement as Octrees.

**Reply by patrikpatrik** [reply]
That's what I figured. Just thought I'd ask. I'm definitely going to tweak with the # of solvers and test out in release mode! Cheers~

**Reply by DanielChappuis** [reply]
You should first compile the library (and your code) in Release mode. The performance should be much better. You should never test the performance of code in Debug mode because of all the asserts and compiler optimizations missing it can run quite slowly.

**Reply by patrikpatrik** [reply]
Just a little update. I finally got release mode working and now I went from 200 objects at once to around 1000 objects before fps dropped from 144 to 70.

However, I think the best performance out of all of the default cube, capsule, custom convex mesh cube, custom convex 3-sided pyramid, custom convex mesh 4-sided pyramid was the default sphere! With lowering the iterations and solver I got a whopping 3000 objects colliding on screen before fps dropped from 144 to 70fps.

This is compiled with Visual Studio 2022 + intel i9-12900K + RTX 3090 Founders Edition. Very cool! 👍
